src/System.Buffers.Experimental/System/Buffers/BufferExtensions.cs

```diff
@@ -117,5 +118,107 @@ internal static int IndexOfStraddling(this ReadOnlySpan<byte> first, IReadOnlyMe
 
             return -1;
         }
+
+        static readonly int s_longSize = Vector<ulong>.Count;
+        static readonly int s_byteSize = Vector<byte>.Count;
+
+        public static int IndexOfVectorized(this Span<byte> buffer, byte value)
+        {
+            Debug.Assert(s_longSize == 4 || s_longSize == 2);
+
+            var byteSize = s_byteSize;
+
+            if (buffer.Length < byteSize * 2 || !Vector.IsHardwareAccelerated) return buffer.IndexOf(value);
+
+            Vector<byte> match = new Vector<byte>(value);
+            var vectors = buffer.NonPortableCast<byte, Vector<byte>>();
+            var zero = Vector<byte>.Zero;
+
+            for (int vectorIndex = 0; vectorIndex < vectors.Length; vectorIndex++)
+            {
+                var vector = vectors.GetItem(vectorIndex);
+                var result = Vector.Equals(vector, match);
+                if (result != zero)
+                {
+                    var longer = Vector.AsVectorUInt64(result);
+                    Debug.Assert(s_longSize == 4 || s_longSize == 2);
+
+                    var candidate = longer[0];
+                    if (candidate != 0) return vectorIndex * byteSize + IndexOf(candidate);
```

Comment:
break and continue as inline returns make codegen nasty; and loop for unrolling as above dotnet/coreclr#7912
```csharp
ulong candidate = 0;
int longIndex = 0;
for (; longIndex < Vector<ulong>.Count; longIndex++)
{
    var candidate = longer[longIndex];
    if (candidate == 0) continue;
    break;
}
return 8 * longIndex + vectorIndex * Vector<byte>.Count + IndexOf(candidate);
```